### PR TITLE
Feat add muting rule schedule fields

### PR DESCRIPTION
**Pull request by jaymcgrath**
Work in progress PR for adding support for scheduling muting rules to terraform.

Running into some odd tfstate updates when receiving validation errors when updating a rule.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -6,6 +6,6 @@ require (
 	github.com/hashicorp/terraform-plugin-sdk v1.16.0
 	github.com/newrelic/go-agent/v3 v3.10.0
 	github.com/newrelic/go-insights v1.0.3
-	github.com/newrelic/newrelic-client-go v0.55.8
+	github.com/newrelic/newrelic-client-go v0.56.1
```
Comment (akane0915) [inline go.mod:9]: Looks like the latest is 0.57.1.  Might want to update this again after your PR to remove omitEmpty from the Go client is merged?

File: newrelic/resource_newrelic_alert_muting_rule.go [hunk 2]
```diff
@@ -1,13 +1,77 @@
...
-
+	"fmt"
 	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
 	"github.com/newrelic/newrelic-client-go/pkg/errors"
+	"log"
+	"time"
 )
 
+func validateNaiveDateTime(val interface{}, key string) (warns []string, errs []error) {
```
Comment (akane0915) [inline newrelic/resource_newrelic_alert_muting_rule.go:12]: Since we don't use `warns` anywhere, should this be a `_`?

File: newrelic/resource_newrelic_alert_muting_rule.go [hunk 3]
```diff
@@ -1,13 +1,77 @@
...
+	valueString := val.(string)
+
+	// test conversion to desired format:
+	_, err := time.Parse("2006-01-02T15:04:05", valueString)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("%#v of %#v must be in the format 2006-01-02T15:04:05", key, valueString))
```
Comment (akane0915) [inline newrelic/resource_newrelic_alert_muting_rule.go:18]: 🎉

File: newrelic/resource_newrelic_alert_muting_rule.go [hunk 4]
```diff
@@ -1,13 +1,77 @@
...
+}
+
+func scheduleSchema() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"end_repeat": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Description:   "The datetime stamp when the MutingRule schedule should stop repeating.",
+				ConflictsWith: []string{"schedule.repeat_count"},
```
Comment (akane0915) [inline newrelic/resource_newrelic_alert_muting_rule.go:30]: This is good.

File: newrelic/resource_newrelic_alert_muting_rule.go [hunk 5]
```diff
@@ -1,13 +1,77 @@
...
+				Optional:     true,
+				Description:  "The datetime stamp representing when the MutingRule should end.",
+				ValidateFunc: validateNaiveDateTime,
+			},
+			"repeat": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "The frequency the MutingRule schedule repeats. One of [DAILY, WEEKLY, MONTHLY]",
+				ValidateFunc: validation.StringInSlice([]string{"DAILY", "WEEKLY", "MONTHLY"}, false),
+			},
```
Comment (akane0915) [inline newrelic/resource_newrelic_alert_muting_rule.go:44]: What is the `false` argument doing?  I'm not sure even looking at the function definition.

File: newrelic/resource_newrelic_alert_muting_rule.go [hunk 6]
```diff
@@ -1,13 +1,77 @@
...
+			"start_time": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "The datetime stamp representing when the MutingRule should start.",
+				ValidateFunc: validateNaiveDateTime,
+			},
+			"time_zone": {
+				Type:        schema.TypeString,
+				Required:    true,
```
Comment (akane0915) [inline newrelic/resource_newrelic_alert_muting_rule.go:60]: Does this mean `time_zone` is required on create and update?  I'm trying to remember how we decided to handle it being optional on update in nerdgraph.

File: newrelic/resource_newrelic_alert_muting_rule.go [hunk 7]
```diff
@@ -133,23 +208,25 @@ func resourceNewRelicAlertMutingRuleRead(d *schema.ResourceData, meta interface{
 }
 
 func resourceNewRelicAlertMutingRuleUpdate(d *schema.ResourceData, meta interface{}) error {
-	client := meta.(*ProviderConfig).NewClient
-	updateInput := expandMutingRuleUpdateInput(d)
-
-	log.Printf("[INFO] Updating New Relic One alert muting rule.")
```
Comment (akane0915) [inline newrelic/resource_newrelic_alert_muting_rule.go:139]: Did we mean to remove this log message?

File: newrelic/resource_newrelic_alert_muting_rule.go [hunk 8]
```diff
@@ -133,23 +208,25 @@ func resourceNewRelicAlertMutingRuleRead(d *schema.ResourceData, meta interface{
...
+	updateInput, err := expandMutingRuleUpdateInput(d)
+	if err != nil {
+		return err
+	}
+
 	_, err = client.Alerts.UpdateMutingRule(accountID, mutingRuleID, updateInput)
 	if err != nil {
-		d.SetId("")
-		return nil
+		return err
```
Comment (akane0915) [inline newrelic/resource_newrelic_alert_muting_rule.go:231]: 🎉  yay for showing error messages

File: newrelic/resource_newrelic_alert_muting_rule_test.go [hunk 9]
```diff
@@ -12,14 +14,64 @@ import (
...
+	resourceName := "schedule.0.end_repeat"
+
+	warns, errs := validateNaiveDateTime(invalidDate, resourceName)
+	expectedErrs := []error{errors.New("\"schedule.0.end_repeat\" of \"123456789123456\" must be in the format 2006-01-02T15:04:05")}
+
+	require.Equal(t, []string([]string(nil)), warns)
+	require.Equal(t, expectedErrs, errs)
+
+}
+
```
Comment (akane0915) [inline newrelic/resource_newrelic_alert_muting_rule_test.go:66]: Love all the naive date time validation tests!

File: newrelic/resource_newrelic_alert_muting_rule_test.go [hunk 10]
```diff
@@ -44,6 +96,66 @@ func TestAccNewRelicAlertMutingRule_Basic(t *testing.T) {
...
+					testAccCheckNewRelicAlertMutingRuleExists(resourceName),
+				),
+			},
+			//Test: Import
+			// TODO - determine why this is failing
+			//	{
+			//		ResourceName:      resourceName,
+			//		ImportState:       true,
+			//		ImportStateVerify: true,
+			//	},
```
Comment (akane0915) [inline newrelic/resource_newrelic_alert_muting_rule_test.go]: What is this import test testing?

File: newrelic/resource_newrelic_alert_muting_rule_test.go [hunk 11]
```diff
@@ -44,6 +96,66 @@ func TestAccNewRelicAlertMutingRule_Basic(t *testing.T) {
...
+		},
+	})
+}
+
```
Comment (akane0915) [inline newrelic/resource_newrelic_alert_muting_rule_test.go:191]: I think it might be worth adding a test to create a muting rule with a schedule and then remove null out the schedule entirely on update.

File: newrelic/structures_newrelic_alert_muting_rule.go [hunk 12]
```diff
@@ -16,10 +20,170 @@ func expandMutingRuleCreateInput(d *schema.ResourceData) alerts.MutingRuleCreate
...
+	return createInput, nil
+}
+
+func expandMutingRuleCreateSchedule(cfg map[string]interface{}) (alerts.MutingRuleScheduleCreateInput, error) {
+	schedule := alerts.MutingRuleScheduleCreateInput{}
+
+	if startTime, ok := cfg["start_time"]; ok {
+		rawStartTime := startTime.(string)
+		if rawStartTime != "" {
+			formattedStartTime, err := time.Parse("2006-01-02T15:04:05", rawStartTime)
```
Comment (akane0915) [inline newrelic/structures_newrelic_alert_muting_rule.go:40]: I'm a little confused why we need `time.Parse("2006-01-02T15:04:05", rawStartTime)`.  Haven't we already validated this is in a naive date time format using the `validateNaiveDateTime` function in `resource_newrelic_alert_muting_rule.go`?

File: newrelic/structures_newrelic_alert_muting_rule.go [hunk 13]
```diff
@@ -16,10 +20,170 @@ func expandMutingRuleCreateInput(d *schema.ResourceData) alerts.MutingRuleCreate
...
+				return alerts.MutingRuleScheduleCreateInput{}, err
+			}
+			schedule.EndRepeat = &alerts.NaiveDateTime{Time: formattedEndRepeat}
+		}
+
+	}
+
+	if repeatCount, ok := cfg["repeat_count"]; ok {
+		r := repeatCount.(int)
+		if r > 0 {
```
Comment (akane0915) [inline newrelic/structures_newrelic_alert_muting_rule.go:87]: Do we need this `r >0` check?  Don't we already validate that in `ValidateFunc:  validation.IntAtLeast(1)`?

File: newrelic/structures_newrelic_alert_muting_rule.go [hunk 14]
```diff
@@ -85,22 +257,28 @@ func expandMutingRuleValues(values []interface{}) []string {
 
 func flattenMutingRule(mutingRule *alerts.MutingRule, d *schema.ResourceData) error {
 
-	x, ok := d.GetOk("condition")
+	x := d.Get("condition")
```
Comment (akane0915) [inline newrelic/structures_newrelic_alert_muting_rule.go:267]: Was this change intentional?

File: newrelic/resource_newrelic_alert_muting_rule.go (hunk 2 above)
Comment (jaymcgrath) [reply newrelic/resource_newrelic_alert_muting_rule.go:12]: That's a good question, I'm not sure! It's part of a named return, so I don't know what the convention is for that if you don't care about one of the values. I'll poke around and see what to do about it

File: newrelic/structures_newrelic_alert_muting_rule_test.go [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+func TestExpandScheduleCreate_EmptyFields(t *testing.T) {
+	// similar to Basic, but assert that empty ("") and omitted fields are left out
+
+	t.Parallel()
+	ts, _ := time.Parse("2006-01-02T15:04:05", "2021-01-21T15:30:00")
+	timestamp := alerts.NaiveDateTime{Time: ts}
+	timeZone := "America/Los_Angeles"
+	//repeat := alerts.MutingRuleScheduleRepeatTypes.WEEKLY
```
Comment (akane0915) [inline newrelic/structures_newrelic_alert_muting_rule_test.go]: Remove?

File: newrelic/structures_newrelic_alert_muting_rule_test.go [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+
+	testSchema := &schema.Set{F: schema.HashString}
+	testSchema.Add("MONDAY")
+	testSchema.Add("TUESDAY")
```
Comment (akane0915) [inline newrelic/structures_newrelic_alert_muting_rule_test.go]: i don't think we're using `testSchema` anywhere in this test?

**Review by akane0915 — CHANGES_REQUESTED** [review]
This is looking great! You made a ton of progress this week. 💯   I left a few comments/questions.  Hopefully we can wrap this up early next week!

File: newrelic/resource_newrelic_alert_muting_rule.go (hunk 5 above)
Comment (jaymcgrath) [reply newrelic/resource_newrelic_alert_muting_rule.go:44]: Ah, yeah, false is for the 'ignoreCase' parameter to this function.  At first I thought "yeah, we want to ignore case to be more flexible, since nerdgraph doesn't care," but then I realized that nerdgraph is going to be sending back ALLUPPERCASE values, which is going to cause state drift if they user has things configured that don't match.

There might be something we could do to tell it to ignore case when diffing, but I just went with the simpler route of requiring screaming snake case

File: newrelic/resource_newrelic_alert_muting_rule.go (hunk 6 above)
Comment (jaymcgrath) [reply newrelic/resource_newrelic_alert_muting_rule.go:60]: I think this is correctly handled on update. It's required when you first create a resource via terraform, and then it's only sent on update if it's been changed in the user's config (diffs vs tf state).

So to reiterate: it's required that the user have a timezone set in their schedule config. On update, terraform will only send the timezone if it's changed. If they try to delete it, it won't send a null on the update because it's marked as a required field.

File: newrelic/resource_newrelic_alert_muting_rule_test.go (hunk 10 above)
Comment (jaymcgrath) [reply newrelic/resource_newrelic_alert_muting_rule_test.go]: terraform can import existing resources - say ones that have been defined via the UI, but you want to take over management with via TF. Import queries an existing resource and then adds it to state after diffing it with how you have it represented in the config. 
Basically, you create a config that reflects an existing rule, then from the CLI run `terraform import newrelic_muting_rule.bar 400304:123456` and it checks the config for `bar` vs the identified rule and if they line up, saves it into state

File: newrelic/resource_newrelic_alert_muting_rule_test.go (hunk 11 above)
Comment (jaymcgrath) [reply newrelic/resource_newrelic_alert_muting_rule_test.go:191]: Yeah, I was just thinking this too! I want to create a bunch of wacky state changes across the acceptance tests like we've been doing by hand locally.

File: newrelic/structures_newrelic_alert_muting_rule.go (hunk 12 above)
Comment (jaymcgrath) [reply newrelic/structures_newrelic_alert_muting_rule.go:40]: The validator will catch incorrectly formatted date strings before attempting to prepare a payload.. the validateNaiveDateTime function takes a time.time, converts it and then returns the errors while throwing away the result.
In this case, we do the conversion for real, going from a go time.time to a string.

File: newrelic/structures_newrelic_alert_muting_rule.go (hunk 14 above)
Comment (jaymcgrath) [reply newrelic/structures_newrelic_alert_muting_rule.go:267]: Yeah, we the previous code wasn't ever doing anything with the value of `ok` -- my ide highlighted it as unused. We were passing it into the flatten function and then ignoring it.

File: newrelic/structures_newrelic_alert_muting_rule_test.go (hunk 15 above)
Comment (jaymcgrath) [reply newrelic/structures_newrelic_alert_muting_rule_test.go]: indeed, good catch

File: newrelic/structures_newrelic_alert_muting_rule_test.go (hunk 16 above)
Comment (jaymcgrath) [reply newrelic/structures_newrelic_alert_muting_rule_test.go]: Taking an axe to it now! 🪓

File: newrelic/structures_newrelic_alert_muting_rule.go (hunk 13 above)
Comment (jaymcgrath) [reply newrelic/structures_newrelic_alert_muting_rule.go:87]: This is a good question. I believe this is handling the cases where we don't have repeat set in the config, so converting the nil pointer to an int results in a 0. I want to remove it and see what happens, though. Thanks for pointing this out.

File: newrelic/resource_newrelic_alert_muting_rule.go (hunk 5 above)
Comment (akane0915) [reply newrelic/resource_newrelic_alert_muting_rule.go:44]: I think that's the right call.  Thanks for explaining!

File: newrelic/resource_newrelic_alert_muting_rule.go (hunk 6 above)
Comment (akane0915) [reply newrelic/resource_newrelic_alert_muting_rule.go:60]: Ok, so terraform won't throw a `time_zone is required` error if `time_zone` isn't provided in an update?

File: newrelic/resource_newrelic_alert_muting_rule_test.go (hunk 10 above)
Comment (akane0915) [reply newrelic/resource_newrelic_alert_muting_rule_test.go]: Ah, I see.  Does this test pass if you uncomment it out?  Or is there still implementation left to do in this PR to enable import?

File: newrelic/structures_newrelic_alert_muting_rule.go (hunk 12 above)
Comment (akane0915) [reply newrelic/structures_newrelic_alert_muting_rule.go:40]: 👍 I keep forgetting we provide a `time.time` type that needs to be converted to a `string`.  For some reason when I wrote this question, I was thinking we provide a `string` and `time.Parse` just reformats that `string`.  My bad.  Thanks!

File: newrelic/resource_newrelic_alert_muting_rule.go (hunk 6 above)
Comment (jaymcgrath) [reply newrelic/resource_newrelic_alert_muting_rule.go:60]: Discussed this in pairing! TZ won't be sent on update if it hasn't changed in the TF config vs state (which is what we want)

File: newrelic/resource_newrelic_alert_muting_rule_test.go (hunk 10 above)
Comment (jaymcgrath) [reply newrelic/resource_newrelic_alert_muting_rule_test.go]: We just ran this, and it passed! 🎉

File: newrelic/resource_newrelic_alert_muting_rule.go (hunk 7 above)
Comment (akane0915) [reply newrelic/resource_newrelic_alert_muting_rule.go:139]: Added back in pairing

**Review by akane0915 — APPROVED** [review]
LGTM!!

File: newrelic/resource_newrelic_alert_muting_rule_test.go (hunk 11 above)
Comment (akane0915) [reply newrelic/resource_newrelic_alert_muting_rule_test.go:191]: Done while pairing.
